### DISCUSSION TITLE
Speed of cursor vs. node traversal

### DISCUSSION THREAD

**Discussion by PaulEibensteiner** [tree-sitter/tree-sitter#2018] [Q&A] [answered]
Hi everybody and thanks for this awesome tool!
If I need to traverse part of a tree and need the type of every node in that part, is it faster to use a cursor and call `ts_tree_cursor_current_node` or traverse the nodes directly? Maybe this could also be added in the documentation

**Comment by razzeee** [comment]
As far as I remember cursor is the fastest way, but maybe wait for someone else to confirm :)

**Comment by sogaiu** [comment]
Don't know if this counts, but FWIW:

> You can access every node in a syntax tree using the TSNode APIs [described above](https://tree-sitter.github.io/tree-sitter/using-parsers#retrieving-nodes), but if you need to access a large number of nodes, the fastest way to do so is with a tree cursor. A cursor is a stateful object that allows you to walk a syntax tree with maximum efficiency.

via: https://tree-sitter.github.io/tree-sitter/using-parsers#walking-trees-with-tree-cursors

**Comment by PaulEibensteiner** [comment]
Hey, yes I saw the paragraph in the documentation. To be clear, as I understand it this means that the cursor is faster, if you want to know if nodes exist (possible directly with the return value) or if you want to only get the node's id (only one call to `ts_tree_cursor_current_field_id`). In my case I would have to call `ts_tree_cursor_current_node` *and* `ts_node_type`, so two function calls for *every* node I am visiting. It probably does not matter much anyways, but I was just curios if you can confirm that using a cursor is still faster in this second case.
Thank you for your fast answers btw :)

**Comment by ahlinc** [comment, marked as answer]
The above statement from documentation about that a cursor is a stateful object means that it internally maintains path to a root node on which it was created. This makes cursor faster because `ts_node_parent` always starts from the top of a tree while cursors `goto_parent` operation is `O(1)`. The `ts_node_parent` operation was faster in the past because [it had an internal cache](https://github.com/tree-sitter/tree-sitter/issues/1469#issuecomment-977860939) that was [removed](https://github.com/tree-sitter/tree-sitter/pull/1119) to make tree-sitter trees fully _immutable_ to allow processing the same tree in parallel from multiple threads without any circumstances.
